README.md
# qbdevice C++ API v7 - Example for SoftHand Research
This project is an example of to set up a CMake project using the compiled libraries obtained by the project [qbdevice-api-7.x.x](https://bitbucket.org/qbrobotics/qbdevice-api-7.x.x/src/production/). The project has the following structure:
```
├── include
│   ├── qbrobotics_research_api
│   └── serial
├── libs
└── src
└── CMakeLists.txt
```
- the folder include/qbrobotics_research_api contains the following header files:

  - qbrobotics_research_api.h

  - qbrobotics_research_commands.h

  - qbsofthand_research_api.h

- the folder include/serial must contain the following header file:

  - serial.h - present in the folder qbdevice-api-7.x.x/serial/include

- the folder libs must contain the compiled libraries obtained by compiling the [qbdevice-api-7.x.x](https://bitbucket.org/qbrobotics/qbdevice-api-7.x.x/src/production/) project.

- the folder src contains the main.cpp that exploits qbrobotics API functionalities.

- The CMakeLists.txt has to configure the project to link against the compiled libraries.

The project can be compiled by running the following commands:
```
mkdir build
cd build
cmake ..
make
```
---

NOTE:
This project does not include the compiled libraries. Compile and paste them in __libs__ folder.

---



CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package qbdevice-api-7.x.x
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


1.0.0 (2022-09-08)
------------------
* Compatible with qbdevice-api-7.x.x version 1.1.3


src/hand_controller.cpp
#include <stdlib.h>

#include <iostream>
#include <chrono>
#include <cmath>
#include <ctime>
#include <iomanip>
#include <memory>
#include <set>
#include <serial/serial.h>
#include <qbrobotics_research_api/qbsofthand_research_api.h>

#include "hand_controller.h"

// handler to manage the communication with qbdevices
std::shared_ptr<qbrobotics_research_api::Communication> communication_handler_;
// Communication ports
std::vector<serial::PortInfo> serial_ports_;
std::map<int, std::shared_ptr<qbrobotics_research_api::qbSoftHandLegacyResearch>> soft_hands_;
// IDs of connected devices 
std::vector<qbrobotics_research_api::Communication::ConnectedDeviceInfo> device_ids_;

HandController::HandController()
{
  initialize();
  getInfo();
}
HandController::~HandController()
{
}
void HandController::initialize()
{
   //scanForDevices : 연결된 device가 몇개인지 check, 안에서 open함수는 serial port를 여는 함수
  std::cout << "qbrobotics devices found: " << scanForDevices(1) << std::endl; 
  _first = true;
  _condition = true;
  iter = 0;
  _end = false;
}
int HandController::open(const std::string &serial_port) {
  if (!std::regex_match(serial_port, std::regex("/dev/ttyUSB[[:digit:]]+"))) {
    return -1;
  }
  if(communication_handler_->openSerialPort(serial_port) < 0){
    std::cerr << "Not able to open: " << serial_port << " serial port";
    return -1;
  }
  std::cout << "Opened: " << serial_port << " serial port"<< std::endl;
  return 0;
}

// Scan ports for qbrobotics devices
int HandController::scanForDevices(const int &max_repeats) {
    communication_handler_ = std::make_shared<qbrobotics_research_api::CommunicationLegacy>(); // make shared pointer that handles the communication
  if (communication_handler_->listSerialPorts(serial_ports_) < 0) {
      std::cerr << "[scanForDevices] no serial ports found" << std::endl; 
      return -1;
  }
  int qbrobotics_devices_found = 0;
  for(auto &serial_port:serial_ports_){ // scan and open all the serial port
    int failures = 0;
    while (failures <= max_repeats) {
      if (open(serial_port.serial_port) != 0) {
        failures++;
        continue;
      }
      break;
    }
    if (failures >= max_repeats) {
      continue;
    }

    if (communication_handler_->listConnectedDevices(serial_port.serial_port, device_ids_) >= 0) { // retrieved at least a qbrobotics device
      for(auto &device_id:device_ids_) {
        if (device_id.id == 120 || device_id.id == 0) {
          std::cout << "Not valid device retrieved!" << std::endl;
          continue;  // ID 120 is reserved, ID 0 is for sure an error
        }
        soft_hands_.insert(std::make_pair(static_cast<int>(device_id.id), std::make_shared<qbrobotics_research_api::qbSoftHandLegacyResearch>(communication_handler_, "dev", serial_port.serial_port, device_id.id)));
        qbrobotics_devices_found++;
      }
      if (qbrobotics_devices_found == 0) {
        std::cerr << "[scanForDevices] no qbrobotics devices found" << std::endl; 
      }
    }
  //   std::cout << "auto &serial_port : "<< &serial_port<< std::endl;
  // std::cout << "qbrobotics_devices_found : "<<qbrobotics_devices_found<< std::endl;
  }
  return qbrobotics_devices_found;
}


void HandController::getInfo()
{
  soft_hands_.at(1)->getInfo(INFO_ALL, info_string); // 연결된 device의 정보 read
  soft_hands_.at(2)->getInfo(INFO_ALL, info_string); // 연결된 device의 정보 read
  std::cout << info_string << std::endl << "----" << std::endl;

  std::cout << "[getControlReferences()] ";                                     //getControlReferences()
  soft_hands_.at(1)->getControlReferences(control_references);
  soft_hands_.at(2)->getControlReferences(control_references1);
  std::cout <<"id1 : ";
  for (auto &reference:control_references){
    std::cout << reference << " ";
  }
  std::cout <<"/ id2 : ";
  for (auto &reference1:control_references1){
    std::cout << reference1 << " ";
  }
  std::cout << "\n----"<< std::endl;

  std::cout << "[getCurrents()] ";                                              //getCurrents()
  soft_hands_.at(1)->getCurrents(currents);
  soft_hands_.at(2)->getCurrents(currents1);
  std::cout <<"id1 : ";
  for (auto &current:currents){
    std::cout << current << " ";
  }
  std::cout <<"/ id2 : ";
  for (auto &currents1:currents1){
    std::cout << currents1 << " ";
  }
  std::cout << "\n----"<< std::endl;

  std::cout << "[getCurrentsAndPositions()] ";                                  //getCurrentsAndPositions()
  soft_hands_.at(1)->getCurrentsAndPositions(currents, positions);
  soft_hands_.at(2)->getCurrentsAndPositions(currents1, positions1);
  std::cout << "id1 : [current] ";
  for (auto &current:currents){
    std::cout <<current << " ";
  }
  std::cout << "[position] ";
  for (auto &position:positions){
    std::cout << position << " ";
  }
  std::cout << "/ id2 : [current] ";
  for (auto &current1:currents1){
    std::cout << current1 << " ";
  }
  std::cout << "[position] ";
  for (auto &position1:positions1){
    std::cout << position1 << " ";
  }
  std::cout << "\n----"<< std::endl;

  std::cout << "[getPositions()] ";                                             //getPositions()
  soft_hands_.at(1)->getPositions(positions);
  soft_hands_.at(2)->getPositions(positions1);
  std::cout <<"id1 : ";
  for (auto &position:positions){
    std::cout << position << " ";
  }
  std::cout <<"/ id2 : ";
  for (auto &position1:positions1){
    std::cout << position1 << " ";
  }
  std::cout << "\n----"<< std::endl;

  std::cout << "[getVelocities()] ";                                            //getVelocities()
  soft_hands_.at(1)->getVelocities(velocities);
  soft_hands_.at(2)->getVelocities(velocities1);
  std::cout <<"id1 : ";
  for (auto &velocity:velocities){
    std::cout << velocity << " ";
  }
  std::cout <<"/ id2 : ";
  for (auto &velocity1:velocities1){
    std::cout << velocity1 << " ";
  }
  std::cout << "\n----"<< std::endl;

  std::cout << "[getAccelerations()] ";                                         //getAccelerations()
  soft_hands_.at(1)->getAccelerations(accelerations);
  soft_hands_.at(2)->getAccelerations(accelerations1);
  std::cout <<"id1 : ";
  for (auto &acceleration:accelerations){
    std::cout << acceleration << " ";
  }
  std::cout <<"/ id2 : ";
  for (auto &acceleration1:accelerations1){
    std::cout << acceleration1 << " ";
  }
  std::cout << "\n----"<< std::endl;

  std::cout << "[setMotorStates()] ";                                           //setMotorStates()
  bool activate = true;
  if (soft_hands_.at(1)->setMotorStates(activate) == 0 && soft_hands_.at(2)->setMotorStates(activate) == 0){
    std::cout << "Motors are active";
  } 
  else {
    std::cout << "Something went wrong while activating motors";
  }
  std::cout << "\n----"<< std::endl;
}
void HandController::controller()
{ 
    auto start = std::chrono::high_resolution_clock::now();
    std::cout << "[setControlReferences()] ";                                     //setControlReferences()
    control_references = {_id1_pos};
    control_references1 = {_id2_pos};
    soft_hands_.at(1)->setControlReferences(control_references);
    soft_hands_.at(2)->setControlReferences(control_references1);
    for (auto &control_reference:control_references){
      std::cout << "id1 : "<<control_reference << " ";
    }
    for (auto &control_reference1:control_references1){
      std::cout << "/ id2 : "<<control_reference1 << " ";
    }
    std::cout << "\n----"<< std::endl;

    std::cout << "[getControlReferences()] ";                                     //getControlReferences()
    soft_hands_.at(1)->getControlReferences(control_references);
    soft_hands_.at(2)->getControlReferences(control_references1);
    std::cout << "id1 : ";
    for (auto &reference:control_references){
      std::cout <<reference << " ";
    }
    std::cout << "/ id2 : ";
    for (auto &reference1:control_references1){
      std::cout << reference1 << " ";
    }
    std::cout << "\n----"<< std::endl;

    std::cout << "[getControlReferences()] ";                                     //getControlReferences()
    soft_hands_.at(1)->getCurrents(currents);
    soft_hands_.at(2)->getCurrents(currents1);
    std::cout << "id1 : ";
    for (auto &current:currents){
      std::cout <<current << " ";
    }
    std::cout << "/ id2 : ";
    for (auto &current1:currents1){
      std::cout << current1 << " ";
    }
    std::cout << "\n----"<< std::endl;
    auto stop = std::chrono::high_resolution_clock::now();
		auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(stop - start);
		std::cout << " during time : "<< duration.count() << std::endl;

    std::this_thread::sleep_for(std::chrono::milliseconds(_time));
  // }
  _condition = true;
}
// void HandController::target(short int pos, int time)
void HandController::target(short int id1_pos, short int id2_pos, int time)
{
  _time = time;
  _id1_pos = id1_pos;
  _id2_pos = id2_pos;
  _condition = false;
}
void HandController::MotionPlan()
{
  time = 2000;
  if(_end == false)
    {
      if (_condition){
        iter +=1;
        if(iter == 1)
        {
          target(19000, 10000, time);
          // _end = true;
        }
        // else if(iter == 2)
        // {
        //   target(0, 0, time);
        //   //  _end = true;
        // }
        // else if(iter == 3)
        // {
        //   target(10000, 19000, time);
        //   // _end = true;
        // }
        // else if(iter == 4)
        // {
        //   target(14000, 5000, time);
        //   // _end = true;
        // }
        // else if(iter == 5)
        // {
        //   target(1000, 10000, time);
        //   // _end = true;
        // }
        // else if(iter == 6)
        // {
        //   target(0, 0, time);
        //   _end = true;
        // }

      }
  }
}
int main() {
  HandController hand; 
  // id 1 : right, id 2 : left
  while(1)
  { 
    hand.MotionPlan();
    while(!hand._condition)
    {
      hand.controller();
    }
  }
  //  Close serial port
  for (auto &port:serial_ports_){
    if(communication_handler_->closeSerialPort(port.serial_port) == 0){
      std::cout << "serial port "<< port.serial_port <<  " closed" << std::endl;
    }
  }
}


include/hand_controller.h
#pragma once
#ifndef __HAND_CONTROLLER_H
#define __HAND_CONTROLLER_H

#include <stdlib.h>
#include <iostream>
#include <chrono>
#include <cmath>
#include <ctime>
#include <iomanip>
#include <memory>
#include <set>
#include <serial/serial.h>
#include <qbrobotics_research_api/qbsofthand_research_api.h>

class HandController
{
    public:
        HandController();
        virtual ~HandController();

        int open(const std::string &serial_port);
        int scanForDevices(const int &max_repeats);

        std::string info_string;
        std::string info_string2;
        std::vector<int16_t> control_references;
        std::vector<int16_t> control_references1;
        std::vector<int16_t> currents;
        std::vector<int16_t> currents1;
        std::vector<int16_t> positions;
        std::vector<int16_t> positions1;
        std::vector<int16_t> velocities;
        std::vector<int16_t> velocities1;
        std::vector<int16_t> accelerations;
        std::vector<int16_t> accelerations1;

        bool _first;

        void controller();
        void getInfo();
        // void target(short int pos, int time);
        void target(short int id1_pos, short int id2_pos, int time);
        void MotionPlan();
        bool _condition;
        int iter;
        bool _end;

    private:
        void initialize();
        int _time;
        short int _id1_pos, _id2_pos; 
        int time;
        int _id;
        
};
#endif


src/hand_controller_backup0911.cpp
#include <stdlib.h>

#include <iostream>
#include <chrono>
#include <cmath>
#include <ctime>
#include <iomanip>
#include <memory>
#include <set>
#include <serial/serial.h>
#include <qbrobotics_research_api/qbsofthand_research_api.h>

#include "hand_controller.h"

// handler to manage the communication with qbdevices
std::shared_ptr<qbrobotics_research_api::Communication> communication_handler_;
// Communication ports
std::vector<serial::PortInfo> serial_ports_;
std::map<int, std::shared_ptr<qbrobotics_research_api::qbSoftHandLegacyResearch>> soft_hands_;
// IDs of connected devices 
std::vector<qbrobotics_research_api::Communication::ConnectedDeviceInfo> device_ids_;

HandController::HandController(int id)
{
  initialize();
  _id = id;
}
HandController::~HandController()
{
}
void HandController::initialize()
{
  _first = true;
  _condition = true;
  iter = 0;
  _end = false;
}
int HandController::open(const std::string &serial_port) {
  if (!std::regex_match(serial_port, std::regex("/dev/ttyUSB[[:digit:]]+"))) {
    return -1;
  }
  if(communication_handler_->openSerialPort(serial_port) < 0){
    std::cerr << "Not able to open: " << serial_port << " serial port";
    return -1;
  }
  std::cout << "Opened: " << serial_port << " serial port"<< std::endl;
  return 0;
}

// Scan ports for qbrobotics devices
int HandController::scanForDevices(const int &max_repeats) {
    communication_handler_ = std::make_shared<qbrobotics_research_api::CommunicationLegacy>(); // make shared pointer that handles the communication
  if (communication_handler_->listSerialPorts(serial_ports_) < 0) {
      std::cerr << "[scanForDevices] no serial ports found" << std::endl; 
      return -1;
  }
  int qbrobotics_devices_found = 0;
  for(auto &serial_port:serial_ports_){ // scan and open all the serial port
    int failures = 0;
    while (failures <= max_repeats) {
      if (open(serial_port.serial_port) != 0) {
        failures++;
        continue;
      }
      break;
    }
    if (failures >= max_repeats) {
      continue;
    }

    if (communication_handler_->listConnectedDevices(serial_port.serial_port, device_ids_) >= 0) { // retrieved at least a qbrobotics device
      for(auto &device_id:device_ids_) {
        if (device_id.id == 120 || device_id.id == 0) {
          std::cout << "Not valid device retrieved!" << std::endl;
          continue;  // ID 120 is reserved, ID 0 is for sure an error
        }
        soft_hands_.insert(std::make_pair(static_cast<int>(device_id.id), std::make_shared<qbrobotics_research_api::qbSoftHandLegacyResearch>(communication_handler_, "dev", serial_port.serial_port, device_id.id)));
        qbrobotics_devices_found++;
      }
      if (qbrobotics_devices_found == 0) {
        std::cerr << "[scanForDevices] no qbrobotics devices found" << std::endl; 
      }
    }
  //   std::cout << "auto &serial_port : "<< &serial_port<< std::endl;
  // std::cout << "qbrobotics_devices_found : "<<qbrobotics_devices_found<< std::endl;
  }
  return qbrobotics_devices_found;
}



void HandController::controller()
{ 
  std::cout << "qbrobotics devices found: " << scanForDevices(2) << std::endl; //scanForDevices : 연결된 device가 몇개인지 check, 안에서 open함수는 serial port를 여는 함수
  std::cout<<"passed"<<std::endl;
  for (auto &id:device_ids_){
    if (id.id == 120 || id.id == 0) {
      std::cout<<"failed"<<std::endl;
      continue;  // ID 120 is reserved, ID 0 is for sure an error
    }
    std::cout << "id : "<<id.id <<std::endl;
    // if (_first){
      _first = false;
      // soft_hands_.at((int)id.id)->getInfo(INFO_ALL, info_string); // 연결된 device의 정보 read
      soft_hands_.at(_id)->getInfo(INFO_ALL, info_string); // 연결된 device의 정보 read
      std::cout << info_string << std::endl << "----" << std::endl;

      std::cout << "[getControlReferences()] ";                                     //getControlReferences()
      soft_hands_.at(_id)->getControlReferences(control_references);
      for (auto &reference:control_references){
        std::cout << reference << " ";
      }
      std::cout << "\n----"<< std::endl;


      std::cout << "[getCurrents()] ";                                              //getCurrents()
      soft_hands_.at(_id)->getCurrents(currents);
      for (auto &current:currents){
        std::cout << current << " ";
      }
      std::cout << "\n----"<< std::endl;

      std::cout << "[getCurrentsAndPositions()] ";                                  //getCurrentsAndPositions()
      soft_hands_.at(_id)->getCurrentsAndPositions(currents, positions);
      for (auto &current:currents){
        std::cout << current << " ";
      }
      for (auto &position:positions){
        std::cout << position << " ";
      }
      std::cout << "\n----"<< std::endl;

      std::cout << "[getPositions()] ";                                             //getPositions()
      soft_hands_.at(_id)->getPositions(positions);
      for (auto &position:positions){
        std::cout << position << " ";
      }
      std::cout << "\n----"<< std::endl;

      std::cout << "[getVelocities()] ";                                            //getVelocities()
      soft_hands_.at(_id)->getVelocities(velocities);
      for (auto &velocity:velocities){
        std::cout << velocity << " ";
      }
      std::cout << "\n----"<< std::endl;

      std::cout << "[getAccelerations()] ";                                         //getAccelerations()
      soft_hands_.at(_id)->getAccelerations(accelerations);
      for (auto &acceleration:accelerations){
        std::cout << acceleration << " ";
      }
      std::cout << "\n----"<< std::endl;

      std::cout << "[setMotorStates()] ";                                           //setMotorStates()
      bool activate = true;
      if (soft_hands_.at(_id)->setMotorStates(activate) == 0){
        std::cout << "Motors are active";
      } else {
        std::cout << "Something went wrong while activating motors";
        break;
      }
      std::cout << "\n----"<< std::endl;
    // }
    std::cout << "[setControlReferences()] ";                                     //setControlReferences()
    control_references = {_pos};
    soft_hands_.at(_id)->setControlReferences(control_references);
    for (auto &control_reference:control_references){
      std::cout << control_reference << " ";
    }
    std::cout << "\n----"<< std::endl;
    std::this_thread::sleep_for(std::chrono::milliseconds(_time));
  }
  _condition = true;
}
void HandController::target(short int pos, int time)
{
  _time = time;
  _pos = pos;
  _condition = false;
}
void HandController::MotionPlan()
{
  time = 1000;
  if(_end == false)
    {
      if (_condition){
        iter +=1;
        if(iter == 1)
        {
          target(19000, time);
          // _end = true;
        }
        else if(iter == 2)
        {
          target(0, time);
          //  _end = true;
        }
        else if(iter == 3)
        {
          target(15000, time);
          // _end = true;
        }
        else if(iter == 4)
        {
          target(0, time);
          _end = true;
        }
        // else if(iter == 5)
        // {
          // target(15000, time);
        // }
        // else if(iter == 6)
        // {
          // target(0, time);
          // _end = true;
        // }
      }
  }
}
int main() {
  HandController hand_R(1);
  HandController hand_L(2);
  while(1)
  { 
    hand_R.MotionPlan();
    hand_L.MotionPlan();
    while(!hand_R._condition)
    {
      hand_R.controller();
      hand_L.controller();
    }
  }
  //  Close serial port
  for (auto &port:serial_ports_){
    if(communication_handler_->closeSerialPort(port.serial_port) == 0){
      std::cout << "serial port "<< port.serial_port <<  " closed" << std::endl;
    }
  }
}


src/main.cpp
/***
 *  Software License Agreement: BSD 3-Clause License
 *
 *  Copyright (c) 2022, qbrobotics®
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
 *  following conditions are met:
 *
 *  * Redistributions of source code must retain the above copyright notice, this list of conditions and the
 *    following disclaimer.
 *
 *  * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
 *    following disclaimer in the documentation and/or other materials provided with the distribution.
 *
 *  * Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
 *    products derived from this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
 *  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 *  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 *  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 *  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 *  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 *  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <stdlib.h>

#include <iostream>
#include <chrono>
#include <cmath>
#include <ctime>
#include <iomanip>
#include <memory>
#include <set>
#include <serial/serial.h>
#include <qbrobotics_research_api/qbsofthand_research_api.h>

// handler to manage the communication with qbdevices
std::shared_ptr<qbrobotics_research_api::Communication> communication_handler_;
// Communication ports
std::vector<serial::PortInfo> serial_ports_;
std::map<int, std::shared_ptr<qbrobotics_research_api::qbSoftHandLegacyResearch>> soft_hands_;
// IDs of connected devices 
std::vector<qbrobotics_research_api::Communication::ConnectedDeviceInfo> device_ids_;

int open(const std::string &serial_port) {
  if (!std::regex_match(serial_port, std::regex("/dev/ttyUSB[[:digit:]]+"))) {
    return -1;
  }
  if(communication_handler_->openSerialPort(serial_port) < 0){
    std::cerr << "Not able to open: " << serial_port << " serial port";
    return -1;
  }
  std::cout << "Opened: " << serial_port << " serial port"<< std::endl;
  return 0;
}

// Scan ports for qbrobotics devices
int scanForDevices(const int &max_repeats) {
    communication_handler_ = std::make_shared<qbrobotics_research_api::CommunicationLegacy>(); // make shared pointer that handles the communication
  if (communication_handler_->listSerialPorts(serial_ports_) < 0) {
      std::cerr << "[scanForDevices] no serial ports found" << std::endl; 
      return -1;
  }
  int qbrobotics_devices_found = 0;
  for(auto &serial_port:serial_ports_){ // scan and open all the serial port
    int failures = 0;
    while (failures <= max_repeats) {
      if (open(serial_port.serial_port) != 0) {
        failures++;
        continue;
      }
      break;
    }
    if (failures >= max_repeats) {
      continue;
    }

    if (communication_handler_->listConnectedDevices(serial_port.serial_port, device_ids_) >= 0) { // retrieved at least a qbrobotics device
      for(auto &device_id:device_ids_) {
        if (device_id.id == 120 || device_id.id == 0) {
          std::cout << "Not valid device retrieved!" << std::endl;
          continue;  // ID 120 is reserved, ID 0 is for sure an error
        }
        soft_hands_.insert(std::make_pair(static_cast<int>(device_id.id), std::make_shared<qbrobotics_research_api::qbSoftHandLegacyResearch>(communication_handler_, "dev", serial_port.serial_port, device_id.id)));
        qbrobotics_devices_found++;
      }
      if (qbrobotics_devices_found == 0) {
        std::cerr << "[scanForDevices] no qbrobotics devices found" << std::endl; 
      }
    }
  }
  return qbrobotics_devices_found;
}


int main() {
  std::cout << "qbrobotics devices found: " << scanForDevices(2) << std::endl; //scanForDevices : 연결된 device가 몇개인지 check, 안에서 open함수는 serial port를 여는 함수
  while(1)
  {
  for (auto &id:device_ids_){
    if (id.id == 120 || id.id == 0) {
      continue;  // ID 120 is reserved, ID 0 is for sure an error
    }
    std::string info_string;
    soft_hands_.at((int)id.id)->getInfo(INFO_ALL, info_string); // 연결된 device의 정보 read
    std::cout << info_string << std::endl << "----" << std::endl;

    std::cout << "[getControlReferences()] ";                                     //getControlReferences()
    std::vector<int16_t> control_references;
    soft_hands_.at((int)id.id)->getControlReferences(control_references);
    for (auto &reference:control_references){
      std::cout << reference << " ";
    }
    std::cout << "\n----"<< std::endl;

    std::cout << "[getCurrents()] ";                                              //getCurrents()
    std::vector<int16_t> currents;
    soft_hands_.at((int)id.id)->getCurrents(currents);
    for (auto &current:currents){
      std::cout << current << " ";
    }
    std::cout << "\n----"<< std::endl;

    std::cout << "[getCurrentsAndPositions()] ";                                  //getCurrentsAndPositions()
    std::vector<int16_t> positions;
    soft_hands_.at((int)id.id)->getCurrentsAndPositions(currents, positions);
    for (auto &current:currents){
      std::cout << current << " ";
    }
    for (auto &position:positions){
      std::cout << position << " ";
    }
    std::cout << "\n----"<< std::endl;

    std::cout << "[getPositions()] ";                                             //getPositions()
    soft_hands_.at((int)id.id)->getPositions(positions);
    for (auto &position:positions){
      std::cout << position << " ";
    }
    std::cout << "\n----"<< std::endl;

    std::cout << "[getVelocities()] ";                                            //getVelocities()
    std::vector<int16_t> velocities;
    soft_hands_.at((int)id.id)->getVelocities(velocities);
    for (auto &velocity:velocities){
      std::cout << velocity << " ";
    }
    std::cout << "\n----"<< std::endl;

    std::cout << "[getAccelerations()] ";                                         //getAccelerations()
    std::vector<int16_t> accelerations;
    soft_hands_.at((int)id.id)->getAccelerations(accelerations);
    for (auto &acceleration:accelerations){
      std::cout << acceleration << " ";
    }
    std::cout << "\n----"<< std::endl;

    std::cout << "[setMotorStates()] ";                                           //setMotorStates()
    bool activate = true;
    if (soft_hands_.at((int)id.id)->setMotorStates(activate) == 0){
      std::cout << "Motors are active";
    } else {
      std::cout << "Something went wrong while activating motors";
      break;
    }
    std::cout << "\n----"<< std::endl;

    std::cout << "[setControlReferences()] ";                                     //setControlReferences()
    control_references = {10000};
    soft_hands_.at((int)id.id)->setControlReferences(control_references);
    for (auto &control_reference:control_references){
      std::cout << control_reference << " ";
    }
    std::cout << "\n----"<< std::endl;
    std::this_thread::sleep_for(std::chrono::milliseconds(2000));

    std::cout << "[setControlReferences()] ";                                     //setControlReferences()
    // control_references = {19000};
    short int test;
    test = 1000;
    control_references = {test};
    soft_hands_.at((int)id.id)->setControlReferences(control_references);
    for (auto &control_reference:control_references){
      std::cout << control_reference << " ";
    }
    std::cout << "\n----"<< std::endl;
    std::this_thread::sleep_for(std::chrono::milliseconds(2000));

    // std::cout << "[getMotorStates()] ";                                           //getMotorStates()
    // soft_hands_.at((int)id.id)->getMotorStates(activate);
    // if(activate){
    //   std::cout << "Motors are active";
    // } else {
    //   std::cout << "Motors are not active";
    //   break;
    // }
    // std::cout << "\n----"<< std::endl;

    // std::cout << "[getParamId()] ";                                               //getParamId()
    // uint8_t device_id;
    // soft_hands_.at((int)id.id)->getParamId(device_id);
    // std::cout << (int)device_id << " ";
    // std::cout << "\n----"<< std::endl;

    // std::cout << "[getParamPositionPID()] ";                                      //getParamPositionPID()
    // std::vector<float> PID;
    // soft_hands_.at((int)id.id)->getParamPositionPID(PID);
    // for (auto &param:PID){
    //   std::cout << param << " ";
    // }
    // std::cout << "\n----"<< std::endl;

    // std::cout << "[getParamCurrentPID()] ";                                      //getParamCurrentPID()
    // soft_hands_.at((int)id.id)->getParamCurrentPID(PID);
    // for (auto &param:PID){
    //   std::cout << param << " ";
    // }
    // std::cout << "\n----"<< std::endl;

    // std::cout << "[getParamStartupActivation()] ";                               //getParamStartupActivation()
    // uint8_t activation;
    // soft_hands_.at((int)id.id)->getParamStartupActivation(activation);
    // std::cout << (int)activation;
    // std::cout << "\n----"<< std::endl;

    // std::cout << "[getParamInputMode()] ";                                       //getParamInputMode()
    // uint8_t input_mode;
    // soft_hands_.at((int)id.id)->getParamInputMode(input_mode);
    // std::cout << (int)input_mode;
    // std::cout << "\n----"<< std::endl;

    // std::cout << "[getParamControlMode()] ";                                     //getParamControlMode()
    // uint8_t control_mode;
    // soft_hands_.at((int)id.id)->getParamControlMode(control_mode);
    // std::cout << (int)control_mode;
    // std::cout << "\n----"<< std::endl;

    // std::cout << "[getParamEncoderResolutions()] ";                              //getParamEncoderResolutions()
    // std::vector<uint8_t> encoder_resolutions;
    // soft_hands_.at((int)id.id)->getParamEncoderResolutions(encoder_resolutions);
    // for (auto &param:encoder_resolutions){
    //   std::cout << (int)param << " ";
    // }
    // std::cout << "\n----"<< std::endl;

    // std::cout << "[getParamEncoderOffsets()] ";                                   //getParamEncoderOffsets()
    // std::vector<int16_t> encoder_offsets;                                         
    // soft_hands_.at((int)id.id)->getParamEncoderOffsets(encoder_offsets);
    // for (auto &offset:encoder_offsets){
    //   std::cout << offset << " ";
    // }
    // std::cout << "\n----"<< std::endl;

    // std::cout << "[getParamEncoderMultipliers()] ";                              //getParamEncoderMultipliers()
    // std::vector<float> encoder_multipliers;
    // soft_hands_.at((int)id.id)->getParamEncoderMultipliers(encoder_multipliers);
    // for (auto &param:encoder_multipliers){
    //   std::cout << param << " ";
    // }
    // std::cout << "\n----"<< std::endl;

    // std::cout << "[getParamUsePositionLimits()] ";                               //getParamUsePositionLimits()
    // uint8_t use_position_limits;
    // soft_hands_.at((int)id.id)->getParamUsePositionLimits(use_position_limits);
    // std::cout << (int)use_position_limits;
    // std::cout << "\n----"<< std::endl;

    // std::cout << "[getParamPositionLimits()] ";                                  //getParamPositionLimits()
    // std::vector<int32_t> position_limits;
    // soft_hands_.at((int)id.id)->getParamPositionLimits(position_limits);
    // for (auto &param:position_limits){
    //   std::cout << (int)param << " ";
    // }
    // std::cout << "\n----"<< std::endl;

    // std::cout << "[getParamPositionMaxSteps()] ";                                //getParamPositionMaxSteps()
    // std::vector<int32_t> position_max_steps;
    // soft_hands_.at((int)id.id)->getParamPositionMaxSteps(position_max_steps);
    // for (auto &param:position_max_steps){
    //   std::cout << (int)param << " ";
    // }
    // std::cout << "\n----"<< std::endl;

    // std::cout << "[getParamCurrentLimit()] ";                                    //getParamCurrentLimit()
    // int16_t current_limit;
    // soft_hands_.at((int)id.id)->getParamCurrentLimit(current_limit);
    // std::cout << (int)current_limit << " ";
    // std::cout << "\n--------------\n"<< std::endl;

    // std::cout << "[getPositions()] ";                                                //getPositions()
    // soft_hands_.at((int)id.id)->getPositions(positions);
    // for (auto &position:positions){
    //   std::cout << position << " ";
    // }
    // std::cout << "\n----"<< std::endl;

    // std::cout << "[setMotorStates()] ";                                              //setMotorStates()
    // // activate = false;
    // activate = true;
    // if (soft_hands_.at((int)id.id)->setMotorStates(activate) == 0){
    //   std::cout << "Motors are not active";
    // } else {
    //   std::cout << "Something went wrong while deactivating motors";
    // }
    // std::cout << "\n----"<< std::endl;
    // std::this_thread::sleep_for(std::chrono::milliseconds(1000));
  }
}

  // // Close serial port
  // for (auto &port:serial_ports_){
  //   if(communication_handler_->closeSerialPort(port.serial_port) == 0){
  //     std::cout << "serial port "<< port.serial_port <<  " closed" << std::endl;
  //   }
  // }
  
}
